### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized perc/MoE model with integer arithmetic

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#315]
Public recipe.train().optimizer(adamw).lr(0.001).seed(18763617977110).epochs(1).log(all).stop(0.0).tf(32).run() on backend:nv0 stalls at compile_model -> NativeModelIr::from_graph -> emit -> compile_native_artifact on the NVIDIA ptx/ptxas branch (recipe.rs ~2852-2896) before any epoch arithmetic. The model (recipe.model().perc(8).exp().qi(5).k.layer(8).moe(1,[layer(8),layer(8)]).prelu().iq(2).s.loss(mae)) lowers perc/layer through lower_project and MoE through lower_moe (5419) with expert routing/select, a path not represented by existing nv0 quantized stall issues (271 residual, 302 perc/LSTM Scan, 308 pool/attention, 293 Bayes/GRU). The loader data/categoric/sharded_csv is ignored. Measured 60s exceeds sub-second expectation for small sharded categorical workload; gap is native artifact compilation, not arithmetic or memory traffic.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `64af3eb522cc4821`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling-small:free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict":"new","issue":0,"title":"NVIDIA nv0 training stalls beyond 60s on quantized perc/MoE model with integer arithmetic","body":"Public recipe.train().optimizer(adamw).lr(0.001).seed(18763617977110).epochs(1).log(all).stop(0.0).tf(32).run() on backend:nv0 stalls at compile_model -> NativeModelIr::from_graph -> emit -> compile_native_artifact on the NVIDIA ptx/ptxas branch (recipe.rs ~2852-2896) before any epoch arithmetic. The model (recipe.model().perc(8).exp().qi(5).k.layer(8).moe(1,[layer(8),layer(8)]).prelu().iq(2).s.loss(mae)) lowers perc/layer through lower_project and MoE through lower_moe (5419) with expert routing/select, a path not represented by existing nv0 quantized stall issues (271 residual, 302 perc/LSTM Scan, 308 pool/attention, 293 Bayes/GRU). The loader data/categoric/sharded_csv is ignored. Measured 60s exceeds sub-second expectation for small sharded categorical workload; gap is native artifact compilation, not arithmetic or memory traffic.","rationale":"Performance packet, backend nv0, stable 60s, composition perc(8).exp().qi(5).k + layer(8) + moe(1,[layer(8),layer(8)]).prelu().iq(2).s + loss(mae). Source inspection confirms earliest failure at compile_native_artifact (2852-2896) before epoch work. lower_moe (5419) introduces select/routing primitives distinct from Scan (302/293) and pure contraction/residual (271) paths, so no existing issue captures this earliest cause. Data loader (sharded_csv), seed, loss, and quantization differences configure the same terminal hang but do not change the implementation path; however the MoE block type materially expands the unreported nv0 quantized native-compilation path, requiring a new issue."}
```

## Deterministic failure packet

kind=performance
id=64af3eb522cc4821
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10130 next:10131 composition:18763617977110
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-nv0-10130-389.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/categoric/sharded_csv")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.perc(8).exp().qi(5).k
		.layer(8)
		.moe(1, [layer(8), layer(8)]).prelu().iq(2).s
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(18763617977110)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.tf(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #425.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #425: Verify current NVIDIA Perc and MoE execution</summary>

Closes #315

This is a verification-only PR. It intentionally changes no tracked files.

## Current public result

Commit: `9098de8711552004bb9eabd84c6d8d9c120d4c58`

Command:

```text
RECIPE_DEBUG=1 ./target/release/recipe --device benji:nv0 /home/nate/.cache/recipe/issues/315/reproduction.rs
```

The exact issue model, data path, split, training settings, and TF32 arithmetic complete successfully from cold native artifacts:

- The 24,000-row training artifact begins with `cache=miss`.
- The 6,000-row validation artifact also begins with `cache=miss`.
- Clang emits NVIDIA LLVM bitcode, and `llc` emits PTX for both artifacts.
- The training epoch completes in 95.587 ms.
- Training MAE is finite at 0.4966, and held-out MAE is finite at 0.5140.
- The full training and evaluation process exits 0 after 24.73 seconds.

The reported 60-second NVIDIA native-artifact stall does not reproduce on current `minimal`. No production-code change is supported by this result.


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.
